src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -245,12 +430,10 @@ public static boolean isWrapperInstance(Object x) {
     }
 
     private static WrapperInstance asWrapperInstance(Object x) {
-        try {
-            if (x != null)
-                return (WrapperInstance) x;
-        } catch (ClassCastException ex) {
-        }
-        throw newIllegalArgumentException("not a wrapper instance");
+        if (x instanceof WrapperInstance wrapperInstance)
```

Comment:
In the previous version, `WrapperInstance` was not needed.  Instead it checks if the class is a  MHProxy class.   Did you run into any issue that you resurrect `WrapperInstance`?    Is it just because of accessing `ensureOriginalLookup`?

Reply:
John Rose argues against using class data for Leyden. Since class data is the only known way to defend against user-manufactured annotations, I switched back to an extra wrapper interface. This approach also requires the Class loading checks since the interface is not unconditionally exported and fails security manager.

Reply:
> This approach also requires the Class loading checks since the interface is not unconditionally exported and fails security manager.

Are you referring to `ClassLoader::checkPackageAccess` change?   As MHProxy implements `sun.invoke.WrapperInstance`, it needs to export `java.base/sun.invoke` to the dynamic module.    The change to `ClassLoader::checkPackageAccess` doesn't make sense for this.

Reply:
Yes. Unfortunately, it fails at https://github.com/openjdk/jdk/blob/bd21e68e679c41c98f39dd2fbd38270ae7d55ed9/src/java.base/share/classes/java/lang/SecurityManager.java#L1317 because a conditionally-exported package is considered a non-(unconditionally-)exported package.

Reply:
If we keep a cache of all MHProxy classes, I think this would not need `sun.invoke.WrapperInstance`.   OTOH, `sun.invoke` does not have any other class except `WrapperInstance` and the `ensureOriginalLookup` method would need to be in an internal class.   So no problem of implementing `sun.invoke.WrapperInstance`.

Reply:
> because a conditionally-exported package is considered a non-(unconditionally-)exported package.

OK.  I need to look into the right solution for this.   The Proxy implementation uses null protection domain to define the proxy class whereas this patch uses the protection domain as the interface to define the MHProxy class.  That's why this issue only occurs in this new implementation.

Reply:
If we can avoid implementing `sun.invoke.WrapperInstance`, this package access check issue would go away.

Reply:
I think we can probably insert a static final field in a wrapper instance class to point to the implemented class and verify with the `PROXY_CLASS_INFOS` ClassValue, much like VarForm field in VarHandle implementations. This should have less reflective impact than the annotation-based approach, and can be cached in a ClassValue as well. Should we cache it though?

Reply:
Is it more straight-forward to keep a `ClassValue<Boolean>` and each wrapper instance class has an entry of `TRUE` value; otherwise, `FALSE`?

Reply:
> Is it more straight-forward to keep a `ClassValue<Boolean>` and each wrapper instance class has an entry of `TRUE` value; otherwise, `FALSE`?

Then we still need to obtain the implemented interface and original method handle information every time they are queried. Having these information (or the method handle providing access) computed early is more convenient.

> If we can avoid implementing sun.invoke.WrapperInstance, this package access check issue would go away. Do you think you can look into it?

Apparently, calling utility methods from `sun.invoke`, like the `ensureOriginalLookup` in this patch, also triggers this check. However, I still think removing the `WrapperInstance` supertype is better, as this makes the code cleaner in reflection.

Reply:
> Apparently, calling utility methods from sun.invoke, like the ensureOriginalLookup in this patch, also triggers this check. 

Yes.  `java.base` can export `sun.invoke` to the dynamic module (it may be worth considering putting the internal interface in a specific package for the dynamic module' use - `sun.invoke` is ok since it currently has one interface but hard to catch when someone adds a new class/interface in `sun.invoke` unintentionally and leak out sensitive API.

Reply:
> Then we still need to obtain the implemented interface and original method handle information every time they are queried. Having these information (or the method handle providing access) computed early is more convenient.

I was thinking the wrapper instance class still implements some private methods to get the implemented interface and original method handle which can be accessed only `java.base` via deep reflection.